Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp

```diff
@@ -885,54 +917,63 @@ bool Storage::shouldStoreBodyAsBlob(const Data& bodyData)
+    assertIsMainThread();
 
     auto& writeOperation = *writeOperationPtr;
-    m_activeWriteOperations.add(WTFMove(writeOperationPtr));
+    auto identifier = writeOperation.identifier();
+    m_activeWriteOperations.add(identifier, WTFMove(writeOperationPtr));
 
+    auto record = writeOperation.record();
     // This was added already when starting the store but filter might have been wiped.
-    addToRecordFilter(writeOperation.record.key);
-
-    backgroundIOQueue().dispatch([this, &writeOperation] {
-        auto recordDirectorPath = recordDirectoryPathForKey(writeOperation.record.key);
-        auto recordPath = recordPathForKey(writeOperation.record.key);
+    addToRecordFilter(record.key);
 
+    backgroundIOQueue().dispatch([this, protectedThis = Ref { *this }, identifier, record = crossThreadCopy(record)]() mutable {
+        auto recordDirectorPath = recordDirectoryPathForKey(record.key);
+        auto recordPath = recordPathForKey(record.key);
         FileSystem::makeAllDirectories(recordDirectorPath);
 
-        ++writeOperation.activeCount;
+        addWriteOperationActivity(identifier);
 
-        bool shouldStoreAsBlob = shouldStoreBodyAsBlob(writeOperation.record.body);
-        auto blob = shouldStoreAsBlob ? storeBodyAsBlob(writeOperation) : std::nullopt;
-
-        auto recordData = encodeRecord(writeOperation.record, blob);
+        bool shouldStoreAsBlob = shouldStoreBodyAsBlob(record.body);
+        auto blob = shouldStoreAsBlob ? storeBodyAsBlob(identifier, record) : std::nullopt;
+        auto recordData = encodeRecord(record, blob);
 
         auto channel = IOChannel::open(WTFMove(recordPath), IOChannel::Type::Create);
         size_t recordSize = recordData.size();
-        channel->write(0, recordData, WorkQueue::main(), [this, &writeOperation, recordSize](int error) {
+        channel->write(0, recordData, WorkQueue::main(), [this, protectedThis = WTFMove(protectedThis), identifier, recordSize](int error) {
             // On error the entry still stays in the contents filter until next synchronization.
             m_approximateRecordsSize += recordSize;
-            finishWriteOperation(writeOperation, error);
+            finishWriteOperationActivity(identifier, error);
 
             LOG(NetworkCacheStorage, "(NetworkProcess) write complete error=%d", error);
         });
     });
 }
 
-void Storage::finishWriteOperation(WriteOperation& writeOperation, int error)
+void Storage::addWriteOperationActivity(WriteOperationIdentifier identifier)
 {
-    ASSERT(RunLoop::isMain());
-    ASSERT(writeOperation.activeCount);
-    ASSERT(m_activeWriteOperations.contains(&writeOperation));
+    Locker locker { m_activitiesLock };
+    m_writeOperationActivities.add(identifier);
+}
 
-    if (--writeOperation.activeCount)
-        return;
+bool Storage::removeWriteOperationActivity(WriteOperationIdentifier identifier)
+{
+    Locker locker { m_activitiesLock };
+    ASSERT(m_writeOperationActivities.contains(identifier));
```

Comment:
This seems fishy. Seems this function is always expected to return true, so why are we returning a boolean at all? And if we don't need that boolean, then why do we need `m_writeOperationActivities` at all?

Reply:
m_writeOperationActivities is a `HashCountedSet`, so it tracks of WriteOperation with activity and the count of activities.
`m_writeOperationActivities.contains(identifier)` is to ensure the WriteOperation has activity.
`m_writeOperationActivities.remove(identifier);`  is decrementing count by 1. It returns true if the count drops to 0.